### PR TITLE
Move debugging section of OpenMPI faq to new section in user documentation

### PR DESCRIPTION
**Pull request by drwootton**
Move debugging topics in Open MPI documentation from the FAQ section into a new top level section in the document.

Cleanup some text in debugging topics

I created separate files for each of the individual topics in the original debugging section. I attempted to clean up headings so they look more like reference material than FAQ pages, along with some cleanup of text to make the style closer to a user reference document.

I noticed a reference to x86 32 bit registers in the memcheck section and references to xhost in the serial debugging section.

Both of these feel like they could be outdated, but I assume they are still correct and made no intentional technical changes to them.

Once I get text from @rhc54 for PMIx tool interface usage I can add another section for that.

Signed-off-by: David Wootton <dwootton@us.ibm.com>

### CODE REVIEW

File: docs/app-debug/mpir-tools.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+following these instructions.
+
+* Build a current version of Open MPI and install it into a sub-directory of the
+  test directory, for instance, ``/home/shim_test/ompi``. The build should use the
+  versions of PMIX and PRRTE included with Open MPI. The configure
+  ``--with-pmix``, ``-with-prrte``, ``--with-pmix-libdir`` and
+  ``--with-prrte-libdir`` flags should not be used.
+
+* Clone the mpir-to-pmix-guide Github repository into the test directory by
+  running ``git clone git@github.com:openpmix/mpir-to-pmix-guide.git``.
```
Comment (jsquyres) [inline docs/app-debug/mpir-tools.rst:37]: Use the HTTPS URL, not the git: URL.

File: docs/app-debug/mpir-tools.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
+Using MPIR-based tools with Open MPI
```
Comment (jsquyres) [inline docs/app-debug/mpir-tools.rst:1]: This section appears to be all-new.  I have a few top-level thoughts:

1. We should probably explain what MPIR is, and that it has been replaced by the PMIx tool attach thingy blah blah blah (I assume that this is the kind of info you are waiting for from Ralph).  But since so many legacy tools still use MPIR, there's a way to use these tools: the MPIR shim.
2. Have a subsection on what the MPIR shim is (no more than 1-3 sentences, I would think -- let the MPIR shim web site do the bulk of the explaining), and a link out to the MPIR shim web site.
3. Have a subsection on how to build/install Open MPI with the MPIR shim.  
    * In the "validation" section, you have a bullet list of installation points.  It might be ok to replace most of that with a code block with the individual commands in a code block, somewhat like an annotated shell script (vs. a bullet for each command and explanation of that command).
4. Have a subsection about how to validate that the MPIR shim is working (e.g., export the shell variables, compile the test MPI program, run it, attach to it, ... etc.).

File: docs/app-debug/mpir-tools.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+* Create the following test program, ``testprog.c``. The duration of the sleep can
+  be modified to whatever value suspends the test target program long enough
+  for you to attach to running mpirun processes:
+
+.. code-block:: sh
+
+  #include <mpi.h>
+  #include <unistd.h>
+  void main(int argc, char **argv) {
```
Comment (jsquyres) [inline docs/app-debug/mpir-tools.rst:72]: `main()` returns an int.

File: docs/app-debug/debug-options.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+* ``mpi_no_free_handles``: If set to true (any positive value), do not
+  actually free MPI objects when their corresponding MPI "free"
+  function is invoked (e.g., do not free communicators when
+  ``MPI_COMM_FREE`` is invoked).  This can be helpful in tracking down
+  applications that accidentally continue to use MPI handles after
+  they have been freed.
+
+* ``mpi_show_mca_params``: If set to true (any positive value), show a
+  list of all MCA parameters and their values during ``MPI_INIT``.
```
Comment (jjhursey) [inline docs/app-debug/debug-options.rst:40]:
```suggestion
  list of all MCA parameters and their values during :ref:`MPI_Init(3) <mpi_init>`.
```

File: docs/app-debug/debug-options.rst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  with), but it can somewhat slow down the startup of large-scale MPI
+  jobs.
+
+* ``mpi_abort_delay``: If nonzero, print out an identifying message
+  when ``MPI_ABORT`` is invoked showing the hostname and PID of the
+  process that invoked ``MPI_ABORT``, and then delay that many seconds
+  before exiting.  A negative value means to delay indefinitely.  This
+  allows a user to manually come in and attach a debugger when an
+  error occurs.  Remember that the default MPI error handler |mdash|
+  ``MPI_ERRORS_ABORT`` |mdash| invokes ``MPI_ABORT``, so this
```
Comment (jjhursey) [inline docs/app-debug/debug-options.rst:62]: Swap the `MPI_ABORT` to
```
:ref:`MPI_Abort(3) <mpi_abort>`
```

Same with some other references to `MPI_ABORT` later in the file.

File: docs/app-debug/mpir-tools.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+A shim module is available for use with debuggers and tools which rely on the
+`MPIR specification <https://www.mpi-forum.org/docs/mpir-specification-03-01-2018.pdf>`_
+to access application process mappings. This module runs as an intermediate process
+between the debugger or other tool and the ``mpirun`` command to obtain this
+information.
+
+You may need to build this module before you can use it. The source code for this
+module can be downloaded from 
+MPIR to PMIx Shim repository <https://github.com/openpmix/mpir-to-pmix-guide>`_.
```
Comment (jjhursey) [inline docs/app-debug/mpir-tools.rst:17]:
```suggestion
`MPIR to PMIx Shim repository <https://github.com/openpmix/mpir-to-pmix-guide>`_.
```

File: docs/app-debug/mpir-tools.rst (hunk 3 above)
Comment (jjhursey) [reply docs/app-debug/mpir-tools.rst:72]: Additionally add `return 0;` after `MPI_Finalize`

File: docs/app-debug/serial-debug.rst [hunk 7]
```diff
@@ -0,0 +1,174 @@
+Using Serial Debuggers to Debug Open MPI Applications
+=====================================================
+
+Serial debugers, such as ``gdb``, can be used to debug Open MPI
+applications.
+
+This section describes two ways to do this.
```
[the remark this answers is not in the document]
Comment (jjhursey) [reply docs/app-debug/serial-debug.rst:7]: Should be "GNU debugger" not "GNU parallel debugger". Otherwise I like the suggestion.

**Review by jjhursey — CHANGES_REQUESTED** [review]

File: docs/app-debug/debug-options.rst (hunk 4 above)
Comment (jsquyres) [reply docs/app-debug/debug-options.rst:40]: Oops -- I missed this one.  I'd rephrase to say something like "when MPI is initialized" (because it's not just MPI_INIT anymore).

**Comment by drwootton** [comment]
@jsquyres @jjhursey I think I made all the requested changes.

File: docs/app-debug/valgrind.rst [hunk 8]
```diff
@@ -0,0 +1,24 @@
+Using Valgrind to Find Open MPI Application Errors
+==================================================
+
+The `Valgrind <https://valgrind.org/>` utilities may be useful in identify application
+logic errors Howver, there are many situations where Open MPI purposefully does
```
Comment (jjhursey) [inline docs/app-debug/valgrind.rst:5]: Fix a couple of typo, and one formatting issue in the URL.
```suggestion
The `Valgrind <https://valgrind.org>`_ utilities may be useful in identifying application
logic errors. However, there are many situations where Open MPI purposefully does
```

**Review by jjhursey — CHANGES_REQUESTED** [review]

File: docs/app-debug/valgrind.rst (hunk 8 above)
Comment (jjhursey) [reply docs/app-debug/valgrind.rst:5]: I don't see these changes in the latest push.

The `_` after the URL is needed for it to render as a link. Currently it renders as italic text
image

File: docs/app-debug/valgrind.rst [hunk 9]
```diff
@@ -0,0 +1,24 @@
+Using Valgrind to Find Open MPI Application Errors
+==================================================
+
+The `Valgrind <https://valgrind.org/>_` utilities may be useful in identifying application
```
Comment (jjhursey) [inline docs/app-debug/valgrind.rst:4]:
```suggestion
The `Valgrind <https://valgrind.org/>`_ utilities may be useful in identifying application
```

Comment (jjhursey) [reply docs/app-debug/valgrind.rst]: Looks good now.
image

**Review by jjhursey — APPROVED** [review]

**Comment by jjhursey** [comment]
@jsquyres This looks good to me. Do you want to give it a quick re-review before we merge?

**Comment by jsquyres** [comment]
I pushed an extra commit with some suggestions for the MPIR shim page.  If you like those changes (or even just some of them) squash it down into the main commit.

**Comment by drwootton** [comment]
@jsquyres Your changes look good, thanks. Squashed into main commit.

**Review by jsquyres — APPROVED** [review]
❤️

**Comment by jjhursey** [comment]
@drwootton Thanks for the hard work on this. Please cherry-pick over to v5.0.x branch when you have an opportunity to do so.
